### PR TITLE
Rox 13089 Backend changes for syslog extra fields

### PR DESCRIPTION
**Pull request by ksurabhi91**
## Description
This PR includes changes for adding extra fields to syslog integration
Changes made to 
- proto file to add extra field
- syslog.go to add key value pairs in cef format 
- added feature flag 
## Testing Performed
[syslog.txt](https://github.com/stackrox/stackrox/files/10143982/syslog.txt)

Deployed syslog server on GKE and tested alert notification sent on syslog server

attached log output from syslog server

Created a syslog notifier with key value pairs using POST v1/notifier api. Attached notifier to exec pod policy and created policy violation. Logged alert object in central log.  Attached alert object from log in PR

[out_key_val.txt](https://github.com/stackrox/stackrox/files/10037354/out_key_val.txt)

### CODE REVIEW

File: central/notifiers/syslog/syslog.go [hunk 1]
```diff
@@ -204,12 +224,14 @@ func (s *syslog) wrapSyslogUnstructuredData(severity int, timestamp time.Time, m
...
 func (s *syslog) AlertNotify(ctx context.Context, alert *storage.Alert) error {
-	unstructuredData := alertToCEF(alert)
+
+	unstructuredData := alertToCEF(alert, s.Notifier)
 	severity := alertToSyslogSeverityMap[alert.GetPolicy().GetSeverity()]
 	timestamp, err := types.TimestampFromProto(alert.GetTime())
 	if err != nil {
 		return err
 	}
+	log.Infof("alerttocef:%s", unstructuredData)
```
Comment (clickboo) [inline central/notifiers/syslog/syslog.go:234]: remove before merging

File: pkg/features/list.go [hunk 2]
```diff
@@ -51,4 +51,7 @@ var (
 
 	// ResyncDisabled disables the resync behavior of the kubernetes listeners in sensor
 	ResyncDisabled = registerFeature("Disable the re-sync", "ROX_RESYNC_DISABLED", false)
+
+	// SyslogextraFields enables user to add additional key value pairs in syslog alert notification in cef format
+	SyslogextraFields = registerFeature("Enable extra fields for syslog integration", "ROX_SYSLOG_FIELDS", false)
```
Comment (clickboo) [inline pkg/features/list.go:56]: Nit: SyslogExtraFields  and ROX_SYSLOG_EXTRA_FIELDS

File: central/notifiers/syslog/syslog.go [hunk 3]
```diff
@@ -99,6 +101,17 @@ func validateSyslog(syslog *storage.Syslog) error {
 		return errors.Errorf("invalid facility %s must be between 0 and 7", facility.String())
 	}
 
+	errList := errorhelpers.NewErrorList("Key value pair validation")
```
Comment (theencee) [inline central/notifiers/syslog/syslog.go:104]: This should all be behind a feature flag check

File: central/notifiers/syslog/syslog.go [hunk 4]
```diff
@@ -99,6 +101,17 @@ func validateSyslog(syslog *storage.Syslog) error {
...
+
+	for _, f := range syslog.GetExtraFields() {
+		if f.GetKey() == "" || f.GetValue() == "" {
+			errList.AddString("all extra fields must have both a key and a value")
```
Comment (theencee) [inline central/notifiers/syslog/syslog.go:108]: Error list like this isn't particularly useful because it's just going to be exact same string "all extra fields must have both a key and a value" repeated a bunch of times. 

My 2c: loop through the extra fields and immediately return an error (like line 101) if _any_ of them are missing key or value

**Review by theencee — COMMENTED** [review]
Please add a few unit tests (central/notifiers/syslog/syslog_test.go). At least one for `validateSyslog` returning errors if key/value is missing, one for _not_ returning an error, and one for validating the new extension with extra fields.

**Review by md2119 — CHANGES_REQUESTED** [review]
The extra fields must be added to audit messages as well.

File: pkg/features/list.go (hunk 2 above)
Comment (ksurabhi91) [reply pkg/features/list.go]: done

File: central/notifiers/syslog/syslog.go (hunk 3 above)
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog.go]: done

File: central/notifiers/syslog/syslog.go (hunk 4 above)
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog.go]: done

File: central/notifiers/syslog/syslog.go [hunk 5]
```diff
@@ -149,10 +160,10 @@ func joinRoleNames(roles []*storage.UserInfo_Role) string {
 	return strings.Join(roleNames, ",")
 }
 
-func alertToCEF(alert *storage.Alert) string {
+func alertToCEF(alert *storage.Alert, notifier *storage.Notifier) string {
 	// There will be 4-5 different key/value pairs in this message.  Allocate space for 5 because additional
 	// allocations are more expensive than a slightly large list
-	extensionList := make([]string, 0, 5)
+	extensionList := make([]string, 0, 6)
```
Comment (vjwilson) [inline central/notifiers/syslog/syslog.go:166]: nit: We need to update the comment above; it still says we are allocating 5 fields, to allow for the maximum list size.

File: pkg/features/list.go [hunk 6]
```diff
@@ -51,4 +51,7 @@ var (
 
 	// ResyncDisabled disables the resync behavior of the kubernetes listeners in sensor
 	ResyncDisabled = registerFeature("Disable the re-sync", "ROX_RESYNC_DISABLED", false)
+
+	// RoxSyslogextraFields enables user to add additional key value pairs in syslog alert notification in cef format
+	RoxSyslogextraFields = registerFeature("Enable extra fields for syslog integration", "ROX_SYSLOG_EXTRA_FIELDS", false)
```
Comment (vjwilson) [inline pkg/features/list.go:56]: We also need to add this to the FE type check in 
`ui/apps/platform/src/types/featureFlag.ts`
after `line 13`

**Review by vjwilson — CHANGES_REQUESTED** [review]
1. Request to add the flag to the front-end list of allowed flags.
2. Nit about comment

File: central/notifiers/syslog/syslog_test.go [hunk 7]
```diff
@@ -63,6 +66,38 @@ func makeNotifier() *storage.Notifier {
 	}
 }
 
+func makeNotifierExtrafields(empty bool) *storage.Notifier {
+
+	keyVals := make([]*storage.KeyValuePair, 1)
```
Comment (theencee) [inline central/notifiers/syslog/syslog_test.go:71]: [nit] you can declare and initialize a slice in one go:

```
keyVals := []*storage.KeyValuePair{{Key: "foo", Value: "bar"}}
```

File: central/notifiers/syslog/syslog_test.go [hunk 8]
```diff
@@ -63,6 +66,38 @@ func makeNotifier() *storage.Notifier {
...
+	pair1 := &storage.KeyValuePair{
+		Key:   "foo",
+		Value: "bar",
+	}
+
+	if empty {
+		pair1.Key = ""
```
Comment (theencee) [inline central/notifiers/syslog/syslog_test.go:78]: also test what happens when there are zero elements in the list:

```
keyVals := []*storage.KeyValuePair{}
```

File: central/notifiers/syslog/syslog_test.go [hunk 9]
```diff
@@ -104,6 +139,45 @@ func (s *SyslogNotifierTestSuite) TestCEFExtensionFromPairs() {
 	s.Equal(fmt.Sprintf("%s %s", extensionPair1, extensionPair2), extension)
 }
 
+func (s *SyslogNotifierTestSuite) TestValidateSyslogEmptyExtrafields() {
+
```
Comment (theencee) [inline central/notifiers/syslog/syslog_test.go:143]: [nit] extraneous newline (for all the methods below)

File: central/notifiers/syslog/syslog_test.go [hunk 10]
```diff
@@ -104,6 +139,45 @@ func (s *SyslogNotifierTestSuite) TestCEFExtensionFromPairs() {
...
+		s.T().SkipNow()
+	}
+
+	notifier := makeNotifierExtrafields(true)
+	sys := notifier.GetSyslog()
+	e := validateSyslog(sys)
+	s.ErrorContains(e, "all extra fields must have both a key and a value")
+}
+
+func (s *SyslogNotifierTestSuite) TestValidateSyslogExtrafields() {
```
Comment (theencee) [inline central/notifiers/syslog/syslog_test.go:155]: [nit] camel case: `TestValidateSyslogExtraFields`

File: central/notifiers/syslog/syslog_test.go [hunk 11]
```diff
@@ -104,6 +139,45 @@ func (s *SyslogNotifierTestSuite) TestCEFExtensionFromPairs() {
...
+		s.T().SkipNow()
+	}
+
+	notifier := makeNotifierExtrafields(false)
+	sys := notifier.GetSyslog()
+	e := validateSyslog(sys)
+	s.Equal(e, nil)
+}
+
+func (s *SyslogNotifierTestSuite) TestValidateAlerttoCef() {
```
Comment (theencee) [inline central/notifiers/syslog/syslog_test.go:168]: [nit] camelCase: `TestValidateAlertToCEF`

File: central/notifiers/syslog/syslog_test.go [hunk 12]
```diff
@@ -104,6 +139,45 @@ func (s *SyslogNotifierTestSuite) TestCEFExtensionFromPairs() {
...
+	if !features.RoxSyslogExtraFields.Enabled() {
+		s.T().Skip("Skip syslog extra fields tests")
+		s.T().SkipNow()
+	}
+
+	notifier := makeNotifierExtrafields(false)
+	testAlert := fixtures.GetAlert()
+	a := alertToCEF(testAlert, notifier)
+	s.Contains(a, "foo=bar")
+}
```
Comment (theencee) [inline central/notifiers/syslog/syslog_test.go:179]: add test for `auditLogToCEF`

File: central/notifiers/syslog/syslog.go [hunk 13]
```diff
@@ -99,6 +100,16 @@ func validateSyslog(syslog *storage.Syslog) error {
 		return errors.Errorf("invalid facility %s must be between 0 and 7", facility.String())
 	}
 
+	if features.RoxSyslogextraFields.Enabled() {
+
```
[the remark this answers is not in the document]
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog.go:104]: done

File: central/notifiers/syslog/syslog.go [hunk 14]
```diff
@@ -99,6 +100,16 @@ func validateSyslog(syslog *storage.Syslog) error {
 		return errors.Errorf("invalid facility %s must be between 0 and 7", facility.String())
 	}
 
+	if features.RoxSyslogextraFields.Enabled() {
```
[the remark this answers is not in the document]
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog.go]: done

File: central/notifiers/syslog/syslog.go [hunk 15]
```diff
@@ -125,8 +136,8 @@ func newSyslog(notifier *storage.Notifier) (*syslog, error) {
 	}, nil
 }
 
-func auditLogToCEF(auditLog *v1.Audit_Message) string {
-	extensionList := make([]string, 0, 8) // There will be 8 different key/value pairs in this message.
+func auditLogToCEF(auditLog *v1.Audit_Message, notifier *storage.Notifier) string {
+	extensionList := make([]string, 0, 9) // There will be 9 different key/value pairs in this message.
```
[the remark this answers is not in the document]
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog.go:140]: done

File: central/notifiers/syslog/syslog.go [hunk 16]
```diff
@@ -204,12 +229,14 @@ func (s *syslog) wrapSyslogUnstructuredData(severity int, timestamp time.Time, m
 }
 
 func (s *syslog) AlertNotify(ctx context.Context, alert *storage.Alert) error {
-	unstructuredData := alertToCEF(alert)
+
```
[the remark this answers is not in the document]
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog.go:232]: done

File: central/notifiers/syslog/syslog_test.go [hunk 17]
```diff
@@ -47,6 +49,7 @@ func (s *SyslogNotifierTestSuite) makeSyslog(notifier *storage.Notifier) *syslog
 }
 
 func makeNotifier() *storage.Notifier {
+
```
[the remark this answers is not in the document]
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog_test.go:52]: done

File: central/notifiers/syslog/syslog_test.go [hunk 18]
```diff
@@ -63,6 +66,38 @@ func makeNotifier() *storage.Notifier {
 	}
 }
 
+func makeNotifierExtrafields(empty bool) *storage.Notifier {
+
```
[the remark this answers is not in the document]
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog_test.go:70]: done

File: central/notifiers/syslog/syslog_test.go (hunk 9 above)
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog_test.go:143]: done

File: central/notifiers/syslog/syslog_test.go [hunk 19]
```diff
@@ -104,6 +124,91 @@ func (s *SyslogNotifierTestSuite) TestCEFExtensionFromPairs() {
 	s.Equal(fmt.Sprintf("%s %s", extensionPair1, extensionPair2), extension)
 }
 
+func (s *SyslogNotifierTestSuite) TestValidateSyslogEmptyExtrafields() {
```
Comment (keyallis) [inline central/notifiers/syslog/syslog_test.go:127]: Everyone writes tests a little differently so grain of salt, but I've become a fan of what Mandar has done with struct definitions and for loops to allow for easy expansion of other test cases. For an example of what that looks like you can check out this function: https://github.com/stackrox/stackrox/blob/1bce7ff4a2c57948e7658b1b91a627330457980e/tests/resourcecollection_test.go#L162

File: central/notifiers/syslog/syslog.go [hunk 20]
```diff
@@ -242,5 +270,6 @@ func (s *syslog) AuditLoggingEnabled() bool {
 
 func (s *syslog) sendSyslog(severity int, timestamp time.Time, messageID, unstructuredData string) error {
 	syslog := s.wrapSyslogUnstructuredData(severity, timestamp, messageID, unstructuredData)
+	log.Infof("syslog us data %s", unstructuredData)
```
[the remark this answers is not in the document]
Comment (clickboo) [reply central/notifiers/syslog/syslog.go:273]: Is there a reason to be log this at Info level? This is at best a debugf imo.

File: central/notifiers/syslog/syslog_test.go (hunk 19 above)
Comment (ksurabhi91) [reply central/notifiers/syslog/syslog_test.go:127]: In case of collections one method is tested with different inputs. This is a great way to do it
For syslog uni tests, multiple method are tested - alertotcef, audittocef, validate syslog. Thats why I wrote separate tests

File: central/notifiers/syslog/syslog.go [hunk 21]
```diff
@@ -99,6 +100,15 @@ func validateSyslog(syslog *storage.Syslog) error {
 		return errors.Errorf("invalid facility %s must be between 0 and 7", facility.String())
 	}
 
+	if features.RoxSyslogExtraFields.Enabled() {
+		for _, f := range syslog.GetExtraFields() {
+			if f.GetKey() == "" || f.GetValue() == "" {
+				return errors.New("all extra fields must have both a key and a value")
+			}
+		}
+
```
Comment (keyallis) [inline central/notifiers/syslog/syslog.go:109]: nit: Unnecessary newline

**Review by keyallis — APPROVED** [review]
nit but otherwise LGTM

**Review by theencee — APPROVED** [review]

**Review by vjwilson — APPROVED** [review]
prep for Front-end work (feature flag) looks good

File: central/notifiers/syslog/syslog_test.go [hunk 22]
```diff
@@ -104,6 +124,91 @@ func (s *SyslogNotifierTestSuite) TestCEFExtensionFromPairs() {
...
+	if !features.RoxSyslogExtraFields.Enabled() {
+		s.T().Skip("Skip syslog extra fields tests")
+		s.T().SkipNow()
+	}
+
+	keyVals := []*storage.KeyValuePair{{Key: "foo", Value: "bar"}}
+	notifier := makeNotifierExtrafields(keyVals)
+	sys := notifier.GetSyslog()
+	e := validateSyslog(sys)
+	s.Equal(e, nil)
```
Comment (md2119) [inline central/notifiers/syslog/syslog_test.go:169]: `s.NoError(e)`

File: central/notifiers/syslog/syslog_test.go [hunk 23]
```diff
@@ -104,6 +124,91 @@ func (s *SyslogNotifierTestSuite) TestCEFExtensionFromPairs() {
...
+		s.T().Skip("Skip syslog extra fields tests")
+		s.T().SkipNow()
+	}
+
+	keyVals := []*storage.KeyValuePair{}
+	notifier := makeNotifierExtrafields(keyVals)
+
+	sys := notifier.GetSyslog()
+	e := validateSyslog(sys)
+	s.Equal(e, nil)
```
Comment (md2119) [inline central/notifiers/syslog/syslog_test.go:152]: `s.NoError(e)`

**Review by md2119 — APPROVED** [review]
a few minor comments, otherwise lgtm
